build/dockerfiles/entrypoint.sh

```diff
@@ -35,94 +35,110 @@ METAS_DIR="${METAS_DIR:-${DEFAULT_METAS_DIR}}"
+        readarray -t images < <(grep "image:" "${meta}" | sed -r "s;.*image:[[:space:]]*'?\"?([._:a-zA-Z0-9-]*/?[._a-zA-Z0-9-]*/[._a-zA-Z0-9-]*(@sha256)?:?[._a-zA-Z0-9-]*)'?\"?[[:space:]]*;\1;")
+        for image in "${images[@]}"; do
+        separators="${image//[^\/]}"
+        # Warning, keep in mind: image without registry name is it possible case. It's mean, that image comes from private registry, where is we have organization name, but no registry name...
+        digest="${imageMap[${image}]}"
+
+        if [[ -z "${digest}" ]] && [ "${#separators}" == "1" ]; then
+            imageWithDefaultRegistry="docker.io/${image}"
+            digest="${imageMap[${imageWithDefaultRegistry}]}"
+        fi
+
+        if [[ -n "${digest}" ]]; then
+            if [[ ${image} == *":"* ]]; then
+            imageWithoutTag="${image%:*}"
+            tag="${image#*:}"
+            else
+            imageWithoutTag=${image}
+            tag=""
+            fi
+
+            REGEX="([[:space:]]*\"?'?)(${imageWithoutTag}):?(${tag})(\"?'?)"
+            sed -i -E "s|image:${REGEX}|image:\1\2${digest}\4|" "$meta"
+        fi
+        done
+    done
+    fi
+
+    update_container_image_references
+
+    exec "${@}"
 
-# We can't use the `-d` option for readarray because
-# registry.centos.org/centos/httpd-24-centos7 ships with Bash 4.2
-# The below command will fail if any path contains whitespace
-readarray -t metas < <(find "${METAS_DIR}" -name 'meta.yaml')
-for meta in "${metas[@]}"; do
-  echo "Checking meta $meta"
-  # Need to update each field separately in case they are not defined.
-  # Defaults don't work because registry and tags may be different.
-  if [ -n "$REGISTRY" ]; then
-    echo "    Updating image registry to $REGISTRY"
-    sed -i -E "s|image:$IMAGE_REGEX|image:\1${REGISTRY}/\3/\4\5:\6\7|" "$meta"
-  fi
-  if [ -n "$ORGANIZATION" ]; then
-    echo "    Updating image organization to $ORGANIZATION"
-    sed -i -E "s|image:$IMAGE_REGEX|image:\1\2/${ORGANIZATION}/\4\5:\6\7|" "$meta"
-  fi
-  if [ -n "$TAG" ]; then
-    echo "    Updating image tag to $TAG"
-    sed -i -E "s|image:$IMAGE_REGEX|image:\1\2/\3/\4:${TAG}\7|" "$meta"
-  fi
-done
-
-exec "${@}"
+}
+
+function update_container_image_references() {
+
+    # We can't use the `-d` option for readarray because
+    # registry.centos.org/centos/httpd-24-centos7 ships with Bash 4.2
+    # The below command will fail if any path contains whitespace
+    readarray -t metas < <(find "${METAS_DIR}" -name 'meta.yaml')
+    for meta in "${metas[@]}"; do
+    echo "Checking meta $meta"
+    # Need to update each field separately in case they are not defined.
+    # Defaults don't work because registry and tags may be different.
+    if [ -n "$REGISTRY" ]; then
+        echo "    Updating image registry to $REGISTRY"
+        < "$meta" tr '\n' '\r' | sed -E "s|image:$IMAGE_REGEX|image:\1${REGISTRY}/\3/\4\5:\6\7|" |  tr '\r' '\n' > "$meta.tmp" && cat "$meta.tmp" > "$meta" && rm "$meta.tmp"
```

Comment:
This `sed` command will only replace the first occurrence.  Some YAML files contain more than one image.  For example, take a look at `/var/www/html/v3/plugins/eclipse/che-theia/latest/meta.yaml` once you are in the container (and do `export CHE_SIDECAR_CONTAINERS_REGISTRY_URL=myregistry.com` and execute `/usr/bin/entrypoint.sh`:
```yaml
  containers:
    - image: 'myregistry.com/eclipse/che-theia@sha256:69b7d27a9e9a4b46c2734d995456385bb0d7ab1022638d95ddaa5a5919ef43c1'
      env:
        - name: THEIA_PLUGINS
          value: 'local-dir:///plugins'
        - name: HOSTED_PLUGIN_HOSTNAME
          value: 0.0.0.0
        - name: HOSTED_PLUGIN_PORT
          value: '3130'
        - name: THEIA_HOST
          value: 127.0.0.1
      mountSources: true
      memoryLimit: 512M
      volumes:
        - name: plugins
          mountPath: /plugins
        - name: theia-local
          mountPath: /home/theia/.theia
      name: theia-ide
      ports:
        - exposedPort: 3100
        - exposedPort: 3130
        - exposedPort: 13131
        - exposedPort: 13132
        - exposedPort: 13133
    - image: 'quay.io/eclipse/che-machine-exec@sha256:98fdc3f341ed683dc0f07176729c887f2b965bade9c27d16dc0e05f9034e624c'
      command:
        - /go/bin/che-machine-exec
        - '--url'
        - '127.0.0.1:3333'
```
Only the first reference is being replaced.  You could fix this by adding a global switch to your `sed` command:
```suggestion
        < "$meta" tr '\n' '\r' | sed -E "s|image:$IMAGE_REGEX|image:\1${REGISTRY}/\3/\4\5:\6\7|g" |  tr '\r' '\n' > "$meta.tmp" && cat "$meta.tmp" > "$meta" && rm "$meta.tmp"
```

I welcome the addition of test cases since those things should be caught at build time.  But I'm questioning the patch about temporarily replacing newlines (\n) by carriage returns (\r).  To me we are applying a band-aid solution to the problem instead of fixing the cause as I explained [here](https://github.com/eclipse/che/issues/19405#issuecomment-823373738).  I understand that some fields could be long (like `description`) and could be better written as multi-line, but are those YAML really meant to be read by humans?  Changing the YAML generation would avoid having to do tricky translations like we are doing now.  The fact that an `image: ` tag gets written as a multi-line field doesn't make sense to me.

This is my opinion and I'm totally fine if you disagree. 😃 No hard feelings!

Reply:
> I would either use `yq` or drop multiline on generation side rather than using complex sed expressions that may still fail.
> We can easily remove multi-line as everything is generated so I would choose this option
> 
> About quotes we can't remove them as they're required due do the `:` character

@benoitf I would use `yq` but it should be part of a separate improvement issue. this PR is a fix on something that already exist. Would creating a new issue to use yq and prioritize it work for you?

Reply:
Thanks @cccs-eric,
I will add a test case for multiple occurrences.
About generating things differently, it would be less robust and may break at any time if we modify the generator. I would rather use yq but would like to have this part of a separate issue as it is a bit more complex (one more dependency and make sure we don't have regressions)